tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java

```diff
@@ -364,6 +368,168 @@ public void testRsaWithoutCrtParams() throws Exception {
+        assertEquals(
+            ((MGF1ParameterSpec) spec1.getMGFParameters()).getDigestAlgorithm(),
+            ((MGF1ParameterSpec) verifier.getParameters().getParameterSpec(PSSParameterSpec.class).getMGFParameters()).getDigestAlgorithm()
+        );
+
+        // Update message digest, assert that we can't upate params, and that signature verifies.
+        signer.update(MESSAGE);
+        verifier.update(MESSAGE);
+        assertThrows(IllegalStateException.class, () -> signer.setParameter(spec2));
+        assertThrows(IllegalStateException.class, () -> verifier.setParameter(spec2));
+        assertTrue(verifier.verify(signer.sign()));
+
+        // Finally, assert that we can set params after reinit (verify doesn't need explicit re-init)
+        signer.initSign(pair.getPrivate());
+        signer.setParameter(spec2);
+        verifier.setParameter(spec2);
+        assertEquals(
+            spec2.getDigestAlgorithm(),
+            signer.getParameters().getParameterSpec(PSSParameterSpec.class).getDigestAlgorithm()
+        );
+        assertEquals(
+            spec2.getDigestAlgorithm(),
+            verifier.getParameters().getParameterSpec(PSSParameterSpec.class).getDigestAlgorithm()
+        );
+        assertEquals(
+            ((MGF1ParameterSpec) spec2.getMGFParameters()).getDigestAlgorithm(),
+            ((MGF1ParameterSpec) signer.getParameters().getParameterSpec(PSSParameterSpec.class).getMGFParameters()).getDigestAlgorithm()
+        );
+        assertEquals(
+            ((MGF1ParameterSpec) spec2.getMGFParameters()).getDigestAlgorithm(),
+            ((MGF1ParameterSpec) verifier.getParameters().getParameterSpec(PSSParameterSpec.class).getMGFParameters()).getDigestAlgorithm()
+        );
+    }
+
+    @Test
+    void testRsaPSSInitSmallKeyAfterSetParameterLargeSaltThrows() throws Exception {
+        final int minimallySecureKeyLen = 2048;
+        final int smallKeySize = 2048 / 4;
+        final Signature signer = Signature.getInstance("RSASSA-PSS", NATIVE_PROVIDER);
+        final int largeSaltLen = 2 * (smallKeySize / 8);
+        final PSSParameterSpec spec = new PSSParameterSpec("SHA-1", "MGF1", MGF1ParameterSpec.SHA1, largeSaltLen, 1);
+
+        // Set PSS params _before_ initializing, so we don't have access to key length, assuming a default of 2048
+        signer.setParameter(spec);
+
+        // Now, initialize the signature with a smaller key
+        final KeyPairGenerator kg = KeyPairGenerator.getInstance("RSA", AmazonCorrettoCryptoProvider.INSTANCE);
+        kg.initialize(smallKeySize);
+        KeyPair pair = kg.generateKeyPair();
+        signer.initSign(pair.getPrivate());
+
+        // Assert that doFinal (i.e. the native method called in signer.sign()) detects overly large salt len and throws
+        byte[] shortMessage = new byte[] { (byte) 0xDE, (byte) 0xAD, (byte) 0xBE, (byte) 0xEF };
+        signer.update(shortMessage);
+        assertThrows(RuntimeCryptoException.class, () -> signer.sign());
```

Comment:
Is this really the correct exception? We can live with it but I'd prefer to see a more specific exception when we call `initSign`

Reply:
hm, this test case is looking for the native provider to throw on signing finalization, not initialization. we could `catch` [this call](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/EvpSignature.java#L385) in `engineSign` and wrap/rethrow with another exception if you think that's more appropriate than RuntimeCryptoException.

which exception were you thinking? `InvalidAlgorithmParameterException`?

Reply:
Can't we detect this failure the moment that either `setParameters()` or `init()` is called? The moment that the salt is incompatible with the key we should detect this and throw and appropriate exception (relative to the call).

Reply:
we can only detect this in `setParameters()` if `init()` has already been called. for greater compatibility, this implementation does not enforce an order between those two calls being made (neither does [Sun](https://github.com/corretto/corretto-11/blob/f2c5b39cdb2afc9a9101620d14cf34c7448f826d/src/java.base/share/classes/sun/security/rsa/RSAPSSSignature.java#L181-L184) nor [BouncyCastle](https://github.com/bcgit/bc-java/blob/bc3b92f1f0e78b82e2584c5fb4b226a13e7f8b3b/prov/src/main/java/org/bouncycastle/jcajce/provider/asymmetric/rsa/PSSSignatureSpi.java#L259-L270)). from the commit message:

> To increase compatibility accross providers, we allow callers to specify
> null params in calls to setParameter (this is consistent with
> BC, but not with SunRsaSign. **Similarly, we don't require
> `setParameter` to be called on RSASSA-PSS instances before `init*` is
> called**, defaulting to digest algorithms of SHA1 unless otherwise
> specified.

i could put checks in each of `init()` and `setParameters()` and throw when either detects an incompatible size. if this sounds like a good approach, is `InvalidAlgorithmParameterException` the correct exception to throw?

Reply:
You know what, I think the current solution is better. Otherwise you need to update the key and parameters in just the correct order and it only matters once they update or do the final call.

Technically we could throw as soon as either update or final is called _but_ I don't think it's worth it. Keep the exception on the final call for simplicity.

However, we should probably throw a [SignatureException](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/SignatureException.html) in that case.

Reply:
1. Is this only possible on the signing flow and not the verification flow? (I only see the new check on the signing flow.)
2. As `SignatureException` is a checked exception, please declare it on the relevant native methods in Java as well.

Reply:
> 1. Is this only possible on the signing flow and not the verification flow? (I only see the new check on the signing flow.)

I believe this is only possible in the signing flow. The condition this test checks is for when `salt length + digest length + message length` exceeds the maximum total message size (referred to as "EM" in the RFC), which is the size of the modulus. Since such a signature can't be generated, there's nothing to verify.

In local testing, no exception was thrown when i tried to verify an empty signature byte array with the oversized salt length + small key parameter set. Signature.verify() simply returned `false`.

> 2. As `SignatureException` is a checked exception, please declare it on the relevant native methods in Java as well.

Ah, thanks for the reminder. done.